src/main.py

```diff
@@ -1887,8 +1887,9 @@ def _transition_primary_switchover(self):
                 self.zk.delete('%s/%s/op' % (self.zk.MEMBERS_PATH, helpers.get_hostname()))
                 self._attach_to_primary(primary, self.config.getfloat('replica', 'recovery_timeout'))
                 return True
-        # Mark switchover node as failure
-        self.zk.write(self.zk.SWITCHOVER_STATE_PATH, 'replica_timed_out', need_lock=False)
+            logging.warning(f'SWITCHOVER_STATE_PATH ({self.zk.SWITCHOVER_STATE_PATH}) became None, but there is no one, who holds the leader lock.')
+            return False
+        logging.warning(f'SWITCHOVER_STATE_PATH ({self.zk.SWITCHOVER_STATE_PATH}) has value {self.zk.get(self.zk.SWITCHOVER_STATE_PATH)}, but expected to be None in timeout. Hope that the new master is doing well.')
```

Comment:
Actually, if we return False from `_transition_primary_switchover`, we would get into situation, when we are primary, not holding lock, and iteration is about to end.
So, on next iteration we would try to acquire lock and open old primary, thus failing switchover completely; there can be no luck for new primary.

This is not really a bad thing to indicate that with `replica_timed_out` state (similar to `master_timed_out`, which is not used too btw).

Maybe, it is better to solve current problem with "Switchover is impossible now" on the other end: in cli when checking   switchover state?
That would be the same: we just ignore this timed out outcome and try to perform switchover again.

Reply:
The problem with replica_timed_out is possible race: _drop_stale_switchover on next iteration of new master deletes switchover.primary node, then old master marks switchover as replica_timed_out. So, after all, new switchover is impossible.

Reply:
I removed master_timed_out too for more consistency. (but i can't imagine any possible problems/races with this state)

Reply:
> The problem with replica_timed_out is possible race: _drop_stale_switchover on next iteration of new master deletes switchover.primary node, then old master marks switchover as replica_timed_out. So, after all, new switchover is impossible.

Ok, now I finally understand. why this is a problem.
Why don't we fix that race? It is already half-fixed: we don't perform cleanup if we don't have lock: https://github.com/yandex/pgconsul/blob/main/src/main.py#L914
We just need to add lock acquiring before writing to `self.zk.SWITCHOVER_STATE_PATH`, and if this node is missing, just give up.
But since it is not used, we can just avoid writing, like you did here.